Python/bytecodes.c

```diff
@@ -3475,10 +3475,29 @@ dummy_func(
             }
             assert(PyTuple_CheckExact(callargs));
 
-            result = do_call_core(tstate, func, callargs, kwargs, cframe.use_tracing);
-            Py_DECREF(func);
-            Py_DECREF(callargs);
-            Py_XDECREF(kwargs);
+            if (Py_IS_TYPE(func, &PyFunction_Type) &&
+                tstate->interp->eval_frame == NULL &&
+                ((PyFunctionObject *)func)->vectorcall == _PyFunction_Vectorcall) {
+                assert(PyTuple_CheckExact(callargs));
+                Py_ssize_t nargs = PyTuple_GET_SIZE(callargs);
+                int code_flags = ((PyCodeObject *)PyFunction_GET_CODE(func))->co_flags;
+                PyObject *locals = code_flags & CO_OPTIMIZED ? NULL : Py_NewRef(PyFunction_GET_GLOBALS(func));
+
+                _PyInterpreterFrame *new_frame = _PyEvalFramePushAndInit_Ex(tstate,
```

Comment:
I think it would make sense to move the unpacking of the tuple into `_PyEvalFramePushAndInit_Ex`.
```
_PyInterpreterFrame *new_frame = _PyEvalFramePushAndInit_Ex(tstate, func, locals, callargs, kwargs).
```

`CALL_FUNCTION_EX` is quite a large instruction. Pushing more code into `_PyEvalFramePushAndInit_Ex` would keep it smaller.

Reply:
Unpacking which tuple?

If you mean move the code above in, that would break convention with the rest of the code base. The other inline calls do these 
```
                int code_flags = ((PyCodeObject *)PyFunction_GET_CODE(func))->co_flags;
                PyObject *locals = code_flags & CO_OPTIMIZED ? NULL : Py_NewRef(PyFunction_GET_GLOBALS(func));
```
before creating the new frame.